luigi_pipeline/lib/hail_tasks.py

```diff
@@ -0,0 +1,62 @@
+"""
+Tasks for Hail.
+"""
+import hail as hl
+import luigi
+from luigi.contrib import gcs
+
+
+def GCSorLocalTarget(filename):
+    target = gcs.GCSTarget if filename.startswith('gs://') else luigi.LocalTarget
+    return target(filename)
+
+
+class VcfFile(luigi.Task):
+    filename = luigi.Parameter()
+    def output(self):
+       return GCSorLocalTarget(self.filename)
+
+
+class HailMatrixTableTask(luigi.Task):
```

Comment:
I think having a HailMatrixTableTask base class is confusing. This might just be about the name - it's a MatrixTableTask but the code only reads from VCF files (instead of reading in a .mt for example) - and a .vcf only gets read in once at the very beginning in the v01 pipeline.  In general I'd vote for using utility functions or composition instead of inheritance for sharing functionality. 
Also, it might be better to implement the full pipeline first to see what abstractions are most useful.

Reply:
Yeah so I thought about the mt functionality, and figured that until it is needed it shouldn't be implemented. But when it is needed, it's not hard to add. The source paths can include `.mt` and it can do logic from that, or we can create another class and rename this to `HailMatrixTableFromVCFTask`.

I think a hail module can added to supplement when there's code that needs to be provided for non-luigi functionality, but this task and subclassing from this is a nice decomposition so we don't have unnecessary luigi-specific housekeeping functionality that bloat up the seqr tasks, and can easily chain them (in case we have another task that reads from the output mt). E.g. if we need to read from mt and we add that functionality to the task, we can create a seqr sublcass and we automatically inherit all that functionality. Furthermore, it's just python convention and specifically luigi convention (e.g. https://github.com/spotify/luigi/blob/master/luigi/contrib/dataproc.py#L43). Python is a great for scripts, but it's an object oriented language and even we should be using that more (e.g. instead of just naming util functions, we should have classes that encapsulate specific functionality).

Reply:
I'm not against using classes sometimes, especially when something is naturally noun-like and has multiple data fields (for example elasticsearch_client.py). The hail code base I think has a nice balance with a few core classes and [many utilities](https://hail.is/docs/0.2/methods/index.html). I'm generally predisposed [against inheritance](https://javarevisited.blogspot.com/2013/06/why-favor-composition-over-inheritance-java-oops-design.html) except where it's clearly better than utilities/composition of objects. 
How about we revisit in subsequent pipeline PRs?

Reply:
Yeah in the next ones as we have more functionality we'll see whether the separation makes code look cleaner and reusable.

Reply:
Sorry, what class should be renamed `ImportVcfTask`? Do you mean `HailMatrixTableTask`, which currently has what you suggested:

>  I could see a base HailMatrixTableTask class that implements dest_path and output and leaves run for subclasses.

Reply:
> Sorry, what class should be renamed ImportVcfTask? Do you mean HailMatrixTableTask.

Yes. I think that either this class's implementation is too specific for a base class or its name is too generic for a concrete class.

> which currently has what you suggested

I should have said "that *only* implements dest_path and output". If `HailMatrixTableTask` is intended as a base class for all tasks that output a MatrixTable, I think a more abstract implementation would be preferable, like having `run` raise NotImplementedError. For example, would a task that imports an MT from a TSV file inherit from `HailMatrixTableTask`? If so, with the current implementation, then it would have  unused `import_vcf` method.

Reply:
Yeah so the case you suggested in similar to Ben's with `mt` files. I'm of the opinion that implementation should happen when needed but we should structure it so it can be easily modified (e.g. for your case it's very easy to support `tsv` in the existing implementation, or separate it when needed). 

The alternatives would be to
- not subclass, which is what Ben's suggesting, but a clear abstraction is needed right now so we don't have a bunch of unnecessary logic in the seqr tasks and can easily add more seqr classes.
- Make  `HailMatrixTableTask` naming super specific like `HailMatrixTableFromVcfTask`, but I think this is premature.

I think having `run raise NotImplementedError` is a good idea however. Keep in mind that these base classes are there to consolidate hail logic from seqr logic, and both are going to change a lot. So the naming is fuzzy to support the changes as needed, but separating them early on is because of luigi and python convention.

@mike-w-wilson I can show you how to `NotImplementedError` works and help you implement it.